rholang/src/main/rholang/cross-shard.rho

```diff
@@ -0,0 +1,524 @@
+                        doneStore!(false) |
+                        statusOut!(["failure", "Invalid signature."])
+                      } else {
+                        votedSetStore!(votedSet.add(pubKey)) |
+                        new currentCountOut, newCountOut in {
+                          // Increment the candidate's vote count
+                          for (@voteMap <- voteMapStore) {
+                            currentCountOut!(voteMap.get(candidate)) |
+                            for (@currentCount <- currentCountOut) {
+                              if (currentCount == Nil) {
+                                voteMapStore!(voteMap.set(candidate, 1)) |
+                                newCountOut!(1)
+                              } else {
+                                voteMapStore!(voteMap.set(candidate, currentCount + 1)) |
+                                newCountOut!(currentCount + 1)
+                              } |
+                              for (@newCount <- newCountOut) {
+                                // Check quorum size
+                                if (newCount < voteQuorumSize) {
+                                  statusOut!(["success", false]) |
+                                  doneStore!(false)
+                                } else {
+                                  // Remove ballot from the registry so it can be garbage collected
+                                  @[*r, "unregister"]!(unregisterBallotToken, Nil) |
+                                  winnerOut!(candidate) |
+                                  statusOut!(["success", true]) |
+                                  doneStore!(true)
+                                }
+                              }
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        } |
+    
+        // cancel: votes to cancel the ballot instead of waiting for remaining keys to vote
+        //
+        //  bytearray pubKey: the public key to vote with
+        //  name statusOut: send success/failure notification on this
+        //  bytearray sig: signature of [*ballot, *statusOut].toByteArray()
+        contract @[*ballot, "cancel"](@pubKey, statusOut, @sig) = {
+          for (@done <- doneStore; @canceledSet <- canceledSetStore) {
+            if (done) {
+              canceledSetStore!(canceledSet) |
+              doneStore!(done) |
+              statusOut!(["fail", "Already done."])
+            } else {
+              // check that the pubkey is in the list
+              if (not pubKeySet.has(pubKey)) {
+                canceledSetStore!(canceledSet) |
+                doneStore!(done) |
+                statusOut!(["fail", "Invalid key."])
+              } else {
+                // check that it hasn't already cast its vote to cancel
+                if (canceledSet.has(pubKey)) {
```

Comment:
Do we have `has` method? Did you mean `contains` ?

Reply:
Yes, thanks.